src/app/data-model-provider/MetadataTypes.h

```diff
@@ -75,45 +88,198 @@ struct ClusterInfo
+
+private:
+    // Constant used to narrow binary expressions
+    static constexpr uint8_t kAttrQualityBits = 7;
+
+    // Mask used in the narrowing of binary expressions
+    static constexpr uint8_t kAttrQualityMask = ((1 << kAttrQualityBits) - 1);
+
+    struct attribute_entry_mask_t
+    {
+
+        // attribute quality flags
+        //
+        // flags is a uint32_t bitfield of size kAttrQualityBits (aka 7),
+        // in order to accomodate all the different values of
+        // "enum class AttributeQualityFlags".
+        //
+        // Consider that any modification on the declaration of
+        // "enum class AttributeQualityFlags" will affect flags.
+        std::underlying_type_t<AttributeQualityFlags> flags : kAttrQualityBits;
+
+        // read/write access privilege variables
+        //
+        // readPrivilege is a uint8_t bitfield of size kPrivilegeBits (aka 5),
+        // in order to accomodate all the different values of "enum class Privilege".
+        // Same case for writePrivilege.
+        //
+        // Consider that any modification on the declaration of "enum class Privilege"
+        // will affect both readPrivilege and writePrivilege.
+        //
+        // The use of bitfields means that each variable holds an individual
+        // Access::Privilege value, as a bitwise value. This allows us to
+        // handle Access::Privilege information without any bit fiddling.
+        std::underlying_type_t<Access::Privilege> readPrivilege : kPrivilegeBits;
+        std::underlying_type_t<Access::Privilege> writePrivilege : kPrivilegeBits;
+    };
+
+    // Make sure that our various flags storage is never larger than 32 bits,
+    // because we expect these fields to pack together.
+    static_assert(sizeof(attribute_entry_mask_t) <= 4, "Size of attribute_entry_mask_t is not as expected.");
+
+    attribute_entry_mask_t mask;
 };
 
+// Static ASSERT to check size of AttributeEntry
+static_assert(sizeof(AttributeEntry) <= 8, "Size of AttributeEntry is not as expected.");
+
 // Bitmask values for different Command qualities.
 enum class CommandQualityFlags : uint32_t
 {
     kFabricScoped = 0x0001,
     kTimed        = 0x0002, // `T` quality on commands
     kLargeMessage = 0x0004, // `L` quality on commands
+
+    // If you add new items here, remember to change kCmdQualityBits
 };
 
 struct AcceptedCommandEntry
 {
     CommandId commandId;
 
-    // TODO: this can be more compact (use some flags for privilege)
-    //       to make this compact, add a compact enum and make flags/invokePrivilege getters (to still be type safe)
-    BitFlags<CommandQualityFlags> flags;
-    Access::Privilege invokePrivilege = Access::Privilege::kOperate;
+    // Constructor
+
+    _StartBitFieldInit; // Disabling '-Wconversion' & '-Wnarrowing'
+
+    constexpr AcceptedCommandEntry(CommandId id = 0, BitMask<CommandQualityFlags> cmdQualityFlags = BitMask<CommandQualityFlags>(),
```

Comment:
OK.  As a followup, could we please change CheckCommandExistence to return a Variant or std::variant of either AcceptedCommandEntry or Status, and then we don't need to have this default-construction bit....